include/crow/websocket.h

```diff
@@ -157,6 +158,36 @@ namespace crow
                     do_write();
                 });
             }
+            
+            /// Send a binary encoded message in blocking mode.
+            void send_binary_blocking(const std::string& msg, boost::system::error_code& ec) override
+            {
+                auto header = build_header(2, msg.size());
+                std::vector<boost::asio::const_buffer> buffer;
+                buffer.emplace_back(boost::asio::buffer(header));
+                buffer.emplace_back(boost::asio::buffer(msg));
+
+                boost::asio::io_service service;
+                boost::asio::deadline_timer deadline_timer(service);
+                deadline_timer.expires_from_now(boost::posix_time::seconds(3));
+
+                deadline_timer.async_wait([&](const boost::system::error_code& err)
+                {
+                    if(!err)
+                    {
+                        ec = boost::system::errc::make_error_code(boost::system::errc::timed_out);
+                        adaptor_.shutdown_write();
```

Comment:
since the adaptor is being shut down, shouldn't `check_destroy()` or `delete this` be used as well?

Reply:
my thought was that the caller decides whether he wants to close the connection in an error case or not. That is also the reason why &ec has to be provided.

Reply:
well if we're shutting down the adaptor I'm not sure there's anything left to do with the connection object, even for the http connection deleting the object was the course of action taken when an error occurs, the `shutdown` parts were added to it later to make the process cleaner.

Reply:
yes, that is correct but I had a different approach. I am saving all connections in a map. If I would call automatically check_destroy() which internally deletes itself (delete this) then I would run into a segmentation fault next time I tried to use it. In this case I can check the &ec and decide to remove that entry from my map. But does some other user know that he has to do that as well only if a write ran into a timeout?

Reply:
A use case like your would be best solved with the close handler, maybe we can call the close method and have that delete the connection, and have the close handler deal with removing the connection from the map.

The approach I'm trying my best to take with Crow is to make getting something up and running as simple as can be while providing the as close as full access to the protocol internals for the convoluted edge cases that end up occurring.

Reply:
I tried to implement that approach but unfortunatelly I was not able to make it run in my asynchronous state machine.
It looks somehow like this now:
The Map is being locked everytime it is being used.
Every now and then something wants to send data -> map is being locked -> send.
During that send we have the case that it somehow hangs -> deadline timer -> check_destroy() -> onclose()-handler.
The onclose wants to lock the map as it wants to remove the connection but this is not possible as the map is still being locked by the send.
Ok so now I outsource the onclose-task in a separate thread that waits for the send to finish and then locks the map to remove the connection.
Now "delete this" is being called as a final step of check_destroy() in the send task.
send-task returns and in an ideal world my outsourced thread will now recognize that it can lock the map to remove the deleted connection.
However, my asynchronous state machine decides to send another data and is faster than the outsourced thread.
Send() locks the map again and runs into a segfault as the connection had been deleted by itself.

Don't get me wrong. I am fine with your idea to have a very simple library and if you want to have check_destroy() being called in such a case then go for it, however, it did not work for me.

Reply:
Maybe returning a bool or enum or something could be applicable.

For your state machine, maybe you could check using an atomic bool flag (or using a similar approach), after sending data, if something has to be removed from the map.
E. g.: Send -> fails and results in onclose() -> sets flag and queues a deletion in a new thread
But before sending anything new, it checks that flag and resets it.